src/NServiceBus.Extensions.Hosting/ServiceCollectionExtensions.cs

```diff
@@ -14,14 +14,18 @@ public static class ServiceCollectionExtensions
         /// </summary>
         /// <remarks>
         /// Use this extension method with WebHost only.
-        /// For GenericHost or ASP.NET Core version 3 or above use <see cref="HostBuilderExtensions.UseNServiceBus"/>
         /// </remarks>
         public static IServiceCollection AddNServiceBus(this IServiceCollection services, EndpointConfiguration configuration)
         {
             var startableEndpoint = EndpointWithExternallyManagedContainer.Create(configuration, new ServiceCollectionAdapter(services));
+            var hostedService = new NServiceBusHostedService(startableEndpoint);
 
-            services.AddSingleton(_ => startableEndpoint.MessageSession.Value);
-            services.AddSingleton<IHostedService>(serviceProvider => new NServiceBusHostedService(startableEndpoint, serviceProvider));
+            services.AddSingleton<IMessageSession>(_ => new WebHostCompatibleMessageSession(hostedService));
```

Comment:
Is there any risk that we would clash with uniformsession?

https://github.com/Particular/NServiceBus.UniformSession/tree/master/src/UniformSession

would it be better to use the TryAdd methods and throw if already present?

Reply:
UniformSession only registers the `IUniformSession` interface afaik so there should be no registration clash

Reply:
Ah yep. Never mind